### PR TITLE
Implement login() in kagglehub

### PR DESCRIPTION
**Pull request by mohami2000**
added login() function, tests, and associated changes in config.py
https://screenshot.googleplex.com/66ZZ3UcBkVUdGCu.png

### CODE REVIEW

File: src/kagglehub/auth.py [hunk 1]
```diff
@@ -1,2 +1,13 @@
...
 def login():
-    raise NotImplementedError()
+    """Prompt the user for their Kaggle username and API key and save them globally."""
+
+    username = input("Enter your Kaggle username: ")
+    api_key = input("Enter your Kaggle API key: ")
+
+    set_kaggle_credentials(username=username, api_key=api_key)
+
+    print("You are now logged in to Kaggle Hub.")
```
Comment (rosbo) [inline src/kagglehub/auth.py:13]: We use `logging` instead of `print` statement to allow users to turn them off if they want to.

Example: https://github.com/Kaggle/kagglehub/blob/6e830e666fa8288b1853707818a3cbdd7a308f5b/src/kagglehub/clients.py#L58

File: src/kagglehub/auth.py [hunk 2]
```diff
@@ -1,2 +1,13 @@
+import json
+from kagglehub.config import set_kaggle_credentials
+
+
 def login():
-    raise NotImplementedError()
+    """Prompt the user for their Kaggle username and API key and save them globally."""
+
+    username = input("Enter your Kaggle username: ")
```
Comment (rosbo) [inline src/kagglehub/auth.py:8]: This is great for the command-line case.

Does this work in a Jupyter notebook? Should we create a nicer widget like I suggested in the bug for the Jupyter case? Perfectly fine to address in a separate PR.

File: src/kagglehub/config.py [hunk 3]
```diff
@@ -58,6 +60,12 @@ def get_kaggle_api_endpoint() -> str:
 
 
 def get_kaggle_credentials() -> Optional[KaggleApiCredentials]:
+    global _kaggle_credentials
+
+    # Check for credentials in the global variable
+    if _kaggle_credentials.username != None and _kaggle_credentials.key != None:
```
Comment (rosbo) [inline src/kagglehub/config.py:66]: Won't this throw a `KeyError` if `_kaggle_credentials` is None?

Replace with: `if _kaggle_credentials:`

File: tests/fixtures.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import unittest
+from kagglehub.config import set_kaggle_credentials
+
+class BaseTest(unittest.TestCase):
+    def setUp(self):
+        # Reset the global variable before each test
+        set_kaggle_credentials(None, None)
```
Comment (rosbo) [inline tests/fixtures.py:7]: Consider adding a `clear_kaggle_credentials` method which would set `_kaggle_credentials` to `None` instead.

File: tests/test_auth.py [hunk 5]
```diff
@@ -1,9 +1,58 @@
...
+from kagglehub.config import get_kaggle_credentials
+from tests.fixtures import BaseTest
+
 
+class TestAuth(BaseTest):
 
-class TestAuth(unittest.TestCase):
-    def test_login(self):
-        with self.assertRaises(NotImplementedError):
+    def test_login_prompts_for_username(self):
```
Comment (rosbo) [inline tests/test_auth.py:14]: A user should always specify both the username and  the api key. They should not be able to specify only the username.

File: tests/test_auth.py [hunk 6]
```diff
@@ -1,9 +1,58 @@
...
+        # Simulate user input for credentials
+        with mock.patch('builtins.input') as mock_input:
+            mock_input.side_effect = ['lastplacelarry', 'some-key']
+            kagglehub.login()
+
+        # Verify that the global variable contains the updated credentials
+        self.assertEqual('lastplacelarry', get_kaggle_credentials().username)
+        self.assertEqual('some-key', get_kaggle_credentials().key)
+
+    def test_login_prints_success_message(self):
```
Comment (rosbo) [inline tests/test_auth.py:41]: No need to add tests for logs. This makes tests more brittle for little added value.

File: tests/fixtures.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import unittest
+from kagglehub.config import set_kaggle_credentials
+
+class BaseTest(unittest.TestCase):
```
Comment (rosbo) [inline tests/fixtures.py:4]: Thanks for adding a fixture for this. I like it. Should we call this `BaseTestCase` to follow the same naming pattern than the `unittest.TestCase`.

**Review by rosbo — CHANGES_REQUESTED** [review]
Congrats on your first PR!

File: src/kagglehub/config.py (hunk 3 above)
Comment (mohami2000) [reply src/kagglehub/config.py]: i cant just have _kaggle_credentials because it doesnt check the fields inside _kaggle_credentials. i think this should fix both problems

File: src/kagglehub/auth.py (hunk 2 above)
Comment (mohami2000) [reply src/kagglehub/auth.py:11]: yes it does. maybe lets address the nicer widget part in a separate cl? it should be simple.

File: src/kagglehub/auth.py [hunk 8]
```diff
@@ -1,2 +1,16 @@
...
+    api_key = input("Enter your Kaggle API key: ")
+
+    set_kaggle_credentials(username=username, api_key=api_key)
+
+    logger.info("You are now logged in to Kaggle Hub.")
```
Comment (rosbo) [inline src/kagglehub/auth.py:16]: We don't validate that the credentials are valid here. It will fail later when a user makes a download / upload call if the credentials are invalid.

Given that, this message is a bit misleading.

I would replace with:

```
logger.info("Kaggle credentials set.")
```

I think it would be great though to validate the credentials at this step. Can you file a bug, assign it to you, add the [kagglehub hotlist](https://b.corp.google.com/hotlists/5354034) and tackle this in a separate PR. 

Thank you

File: tests/test_auth.py [hunk 9]
```diff
@@ -18,3 +18,11 @@ def test_login_updates_global_credentials(self):
         # Verify that the global variable contains the updated credentials
         self.assertEqual("lastplacelarry", get_kaggle_credentials().username)
         self.assertEqual("some-key", get_kaggle_credentials().key)
+
+    def test_set_kaggle_credentials_raises_error_with_empty_username(self):
+        with self.assertRaises(ValueError):
```
Comment (rosbo) [inline tests/test_auth.py:23]: Can you test e2e from kagglehub.login?

File: src/kagglehub/config.py [hunk 10]
```diff
@@ -127,10 +123,13 @@ def _is_env_var_truthy(env_var_name: str) -> bool:
 
 
 def set_kaggle_credentials(username: str, api_key: str):
+    if not username or not api_key:
```
Comment (rosbo) [inline src/kagglehub/config.py:126]: Can you check whether the username is empty or only whitespace (you can use trim for this)?

And add a test to ensure we throw a value error if the user enters " " (whitespace) to for either of these.

File: src/kagglehub/config.py [hunk 11]
```diff
@@ -61,11 +61,7 @@ def get_kaggle_api_endpoint() -> str:
 
 def get_kaggle_credentials() -> Optional[KaggleApiCredentials]:
     # Check for credentials in the global variable
-    if (
-        _kaggle_credentials is not None
-        and _kaggle_credentials.username is not None
-        and _kaggle_credentials.key is not None
-    ):
+    if _kaggle_credentials:
```
Comment (rosbo) [inline src/kagglehub/config.py:64]: Can you add a test inside the `test_config.py` that test the following:

1. Call set_kaggle_credentials to set valid credentials
1. Call get_kaggle_credentials() & assert username & key
1. Call clear_kaggle_credentials()
1. Call get_kaggle_credentials() & assert value is None.

File: tests/test_auth.py (hunk 9 above)
Comment (rosbo) [reply tests/test_auth.py:23]: Instead of calling `set_kaggle_credentials` here, call:

```
with mock.patch("builtins.input") as mock_input:
            mock_input.side_effect = ["", "some-key"]
            kagglehub.login()
```

File: src/kagglehub/config.py [hunk 12]
```diff
@@ -114,3 +120,19 @@ def _get_kaggle_credentials_folder() -> str:
...
 def _is_env_var_truthy(env_var_name: str) -> bool:
     return env_var_name in os.environ and os.environ[env_var_name].lower() in TRUTHY_VALUES
+
+
+def set_kaggle_credentials(username: str, api_key: str):
+    error_message = "Both username and API key cannot be empty or whitespace"
+    stripped_username = username.strip()
+    stripped_api_key = api_key.strip()
+    if not stripped_username or not stripped_api_key:
+        raise ValueError(error_message)
```
Comment (rosbo) [inline src/kagglehub/config.py:130]: nit: You can define the error_message inside the if, just above the call to `raise`.

File: tests/test_auth.py [hunk 13]
```diff
@@ -1,9 +1,30 @@
...
+            with mock.patch("builtins.input") as mock_input:
+                mock_input.side_effect = ["", "some-key"]
+                kagglehub.login()
+
+    def test_set_kaggle_credentials_raises_error_with_empty_api_key(self):
+        with self.assertRaises(ValueError):
+            set_kaggle_credentials(username="lastplacelarry", api_key="")
```
Comment (rosbo) [inline tests/test_auth.py:30]: Use mock.path("builtins.input") here too.

For completeness, let's add another test case for an empty username & password.

**Review by rosbo — APPROVED** [review]
Only some small nit, otherwise LGTM.
